fix: repeat must not mutate its result object in place (#785)

webstrand opened the pull request:
Fixes #785

This substantially rewrites `repeat` to be more straight-forward with fewer closures, unlike #786 which is more conservative.

webstrand commented on packages/range/src/repeat.ts:
```diff
@@ -23,60 +23,68 @@ export function repeat<T>(
-  const memoLen = createMemo(() => Math.floor(Math.max(times(), 0)));
-  return () => {
-    const len = memoLen();
-    return untrack(() => {
```
I removed this `untrack` call+closure, as far as I can tell it's redundant: `createRoot` already does both Owner and Listener.

webstrand commented:
This is actually kind of broken still, if the array returned from `repeat()()` is mutated, internal state can be corrupted

thetarnav commented:
That’s fine
The same probably happens with mapArray and others
When mutating the result one should copy
Nothing broken with that

webstrand commented:
Is it appropriate to narrow the return type to `readonly T[]` then to signal to users not to mutate the return value?

I've narrowed the return type in the most recent batch of commits, and I've fixed a bug in my implementation and added another test.

I'm not sure if you'd rather have a single commit in the PR or multiple?

thetarnav commented:
readonly arrays are not allowed by JSX types, so I wouldn't do that.
